pkg/experiment/metastore/index/index.go

```diff
@@ -65,28 +71,26 @@ func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 
 type Store interface {
 	CreateBuckets(*bbolt.Tx) error
-	ListPartitions(*bbolt.Tx) ([]*store.Partition, error)
-	LoadShard(*bbolt.Tx, store.PartitionKey, string, uint32) (*store.Shard, error)
+	Partitions(tx *bbolt.Tx) iter.Seq[store.Partition]
+	LoadShard(tx *bbolt.Tx, p store.Partition, tenant string, shard uint32) (*store.Shard, error)
+	DeleteShard(tx *bbolt.Tx, p store.Partition, tenant string, shard uint32) error
 }
 
 type Index struct {
-	logger     log.Logger
-	config     Config
-	store      Store
-	partitions []*store.Partition
-	shards     *shardCache
-	blocks     *blockCache
-	mu         sync.RWMutex
+	logger log.Logger
+	config Config
+	store  Store
+	shards *shardCache
+	blocks *blockCache
 }
 
 func NewIndex(logger log.Logger, s Store, cfg Config) *Index {
 	return &Index{
-		logger:     logger,
-		config:     cfg,
-		store:      s,
-		partitions: make([]*store.Partition, 0),
-		shards:     newShardCache(cfg.ShardCacheSize),
-		blocks:     newBlockCache(cfg.BlockReadCacheSize, cfg.BlockWriteCacheSize),
+		logger: logger,
+		config: cfg,
+		store:  s,
+		shards: newShardCache(cfg.ShardCacheSize, s),
+		blocks: newBlockCache(cfg.BlockReadCacheSize, cfg.BlockWriteCacheSize),
```

Comment:
I finally removed the in-memory partition list – it's much cleaner now. Otherwise, we would eventually violate transaction isolation guarantees by exchanging state through the list